### DISCUSSION TITLE
Nullable-generics interaction doesn't work as expected.

### DISCUSSION THREAD

**Discussion by mlavrent** [dotnet/csharplang#7609] [Q&A] [answered]
I posted [this question](https://stackoverflow.com/questions/77319900/function-returning-nullable-generic-type-not-returning-correct-type) recently about how generic return types and the handling of nullable types is handled and how it seems to be not correct.

The TL;DR is this: if I have a method `static A? Test<A>(A input) => default(A?)` that has a nullable return type, it seems to behave incorrectly for value types such as bool e.g. `Test(true)` has type `bool` instead of type `bool?` (or, more explicitly `Nullable<bool>`). This seems to be a result of the fact that the desugaring of nullable types is different depending on whether the type is a reference type or value type. 

The bug (is it a bug or expected?) lies the case where it's a nullable generic parameter that is unconstrained to be either a reference type (i.e. `where A : class`) or a value type (i.e. `where A : struct`). In this case, it gets [desugared](https://sharplab.io/#v2:D4AQTAjAsAULDEA7ArgG1QQwEaoKYAJdFs9ZYQBmfcaiAdlgG9Z9XqwwW3mY2/aAnAAoAKrgDOAFyGSATslwAafHIUBKNQG4urAL478ByrQBs+AIIB+fGKkAecwD4h5/ADcMqZVgD2P1PjI4rgAcmioaqwAvI4GfEGh4fjWACa4AGYYaNJWkQBc7p7aMLpAA) into the version for reference types, causing the weird behavior when `A` is a a value type such as `bool`.

Are there plans to address this? This behavior seems to be incorrect since the actual type of the expression `Test(true)` is `bool` even though the method's return type is nominally `bool?`.

**Comment by CyrusNajmabadi** [comment, marked as answer]
This is by design.  When T is unconstrained (no `struct` or `class` constraint) then `T?` simply means `"can have its default value"` (*not* `can be null`).  

If you have the `class` constraint then it is known to be a reference type, and can be `null`.  Similarly, if you have the `struct` constraint then `T?` means `Nullable<T>` which can also be null.

> Are there plans to address this? 

No.  This has been how generics have worked in .net and c# for 20 years.   Far too much code has been written that depends on things working in this exact fashion.

**Reply by mlavrent** [reply]
Interesting, this is nice context to know!

**Reply by mlavrent** [reply]
On a related, but orthogonal note, are there plans to add Haskell-style `Option`/`Maybe` types to the stdlib?

**Reply by HaloFour** [reply]
See: https://github.com/dotnet/csharplang/issues/113

There's been a bit of movement in this space recently, mostly explorations:

https://github.com/dotnet/csharplang/issues/7016
https://github.com/dotnet/csharplang/issues/7544

It's unlikely that the BCL team would adopt any form of common discriminated union like `Maybe<T>` or `Result<T>` without a good story in the language.

**Reply by seldoff** [reply]
> simply means "can have its default value" 

Is this true, though? E.g., the following prints `True`.
```
public class Test
{
    public static T1 Foo<T1>() where T1 : new() => new();
    public static void Bar() => Console.WriteLine(Foo<int>() == default);
}
```

This means the only constraint T has over T? for value types in this scenario is that you can't use `default` explicitly.

**Reply by CyrusNajmabadi** [reply]
> Is this true, though?

Yes.  It is true.

> E.g., the following prints True.

Yes.  Why wouldn't it?   You are calling `Foo<int>()`.  So that's calling the method `public static int Foo<int>()`.  Calling that method returns `new int()` which is 0.  This is then compared to `default`.  Since the left side of the `==` is `int`, the right side is inferred to be as well.   So you get `Foo<int>() == default(int)`. This is a check that `0 == 0`, which is true.
